### ISSUE TITLE
limit and skip issue 

### ISSUE THREAD

**Issue by martingg88** [orientechnologies/orientdb#4985]
I should expected to have the result here, but i get empty . 
I have this issue for version 2.1.1

SELECT EXPAND(inV()) FROM list WHERE out = #16:11710 limit 8 skip 8
return result set.

SELECT EXPAND(inV()) FROM list WHERE out = #16:11710 ORDER BY (ctime DESC) LIMIT 8 SKIP 7
return result with order clause

SELECT EXPAND(inV()) FROM list WHERE out = #16:11710 ORDER BY (ctime DESC) LIMIT 8 SKIP 8
return empty with order clause

**Comment by luigidellaquila** [comment]
fixed in 2.1.x and develop branches.
Please confirm that it also fixes your use case

**Comment by martingg88** [comment]
yes..it introduce another issue for 2.1.4. 

case 1 as follows work in correct way. 

SELECT inV().@rid FROM list WHERE out = #16:0 
ORDER BY (ctime DESC) LIMIT 8 SKIP 0

SELECT inV().@rid FROM list WHERE out = #16:0 
ORDER BY (ctime DESC) LIMIT 8 SKIP 8

case 2 does't work at all

1. SELECT expand(inV()) FROM list WHERE out = #16:0 
ORDER BY (ctime DESC) LIMIT 8 SKIP 0
--return first 8 records. 

2. SELECT expand(inV()) FROM list 
WHERE out = #16:0 
ORDER BY (ctime DESC) LIMIT 8 SKIP 8
- return another 8 records but these records is as same as with case 2. 1 above (etc: same with first 8 records). weird and faint.

**Comment by martingg88** [comment]
expand screw up order issue. 

1. select expand(in) from (SELECT FROM list WHERE out = #16:0 
ORDER BY (in.ctime DESC) LIMIT 8 SKIP 0)
this  work as expected. 

2. SELECT expand(in)  FROM list WHERE out = #16:0 
ORDER BY (in.ctime DESC) LIMIT 8 SKIP 0

this mess up order for returning result when added expand function directly in query. 

any idea?

**Comment by luigidellaquila** [comment]
Hi @martingg88 


I have some trouble understanding your queries, because I do not have the full picture of your domain, but I guess the problem arises from two factors:
- ORDER BY, SKIP and LIMIT are applied after the expand, are you taking it into consideration?
- actually there was an issue that made this behavior inconsistent in some cases, I fixed it in 2.1.x branch, could you try it?

Thanks

Luigi

**Comment by martingg88** [comment]
i expect ORDER BY, SKIP and LIMIT should happen before the expand, and which mean they should applied to target itself. isn't it?

yes..i have meet  up these issue for 2.1.4 version.

**Comment by martingg88** [comment]
any update?

**Comment by luigidellaquila** [comment]
Hi @martingg88 

not actually, they are all applied after the expand(). 
If you want to apply them before, you just have to use a subquery with ORDER BY, SKIP, LIMIT and do the expand() in the outer query.
Anyway, it's just a convention, if it worked the other way, you would also be able to reproduce the two different behaviors with different subqueries.
I think it's not clear in the docs, I'm going to check.

Thanks

Luigi

**Comment by martingg88** [comment]
will it affect the performance if i need the subquery to get expected result?

**Comment by luigidellaquila** [comment]
Hi @martingg88 

subqueries have always a little cost in performance, but in this case it should not be significant

Thanks

Luigi
